contrib/epee/src/wipeable_string.cpp

```diff
@@ -0,0 +1,142 @@
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+// 
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include <string.h>
+#include "misc_log_ex.h"
+#include "wipeable_string.h"
+
+namespace epee
+{
+
+void *(*wipeable_string::wipefunc)(void*, size_t) = NULL;
+
+wipeable_string::wipeable_string(const wipeable_string &other)
+{
+  buffer = other.buffer;
+}
+
+wipeable_string::wipeable_string(wipeable_string &&other)
+{
+  if (&other == this)
+    return;
+  buffer = std::move(other.buffer);
+}
+
+wipeable_string::wipeable_string(const std::string &other)
+{
+  grow(other.size());
+  memcpy(buffer.data(), other.c_str(), size());
+}
+
+wipeable_string::wipeable_string(std::string &&other)
+{
+  CHECK_AND_ASSERT_THROW_MES(wipefunc, "wipefunc is not set");
+  grow(other.size());
+  memcpy(buffer.data(), other.c_str(), size());
+  if (!other.empty())
+    wipefunc(&other[0], other.size()); // we're kinda left with this again aren't we
+}
+
+wipeable_string::wipeable_string(const char *s): wipeable_string(std::string(s))
+{
+}
+
+wipeable_string::~wipeable_string()
+{
+  wipe();
+}
+
+void wipeable_string::wipe()
+{
+  CHECK_AND_ASSERT_THROW_MES(wipefunc, "wipefunc is not set");
+  wipefunc(buffer.data(), buffer.size() * sizeof(char));
+}
+
+void wipeable_string::grow(size_t sz, size_t reserved)
+{
+  CHECK_AND_ASSERT_THROW_MES(wipefunc, "wipefunc is not set");
+  if (reserved == 0)
+    reserved = sz;
+  CHECK_AND_ASSERT_THROW_MES(reserved >= sz, "reserved < sz");
+  if (reserved <= buffer.capacity())
+    return;
+  size_t old_sz = buffer.size();
+  char *tmp = new char[old_sz];
+  memcpy(tmp, buffer.data(), old_sz * sizeof(char));
+  wipefunc(buffer.data(), old_sz * sizeof(char));
+  buffer.reserve(reserved);
+  buffer.resize(sz);
+  memcpy(buffer.data(), tmp, sz * sizeof(char));
+  wipefunc(tmp, old_sz * sizeof(char));
+  delete[] tmp;
+}
+
+void wipeable_string::push_back(char c)
```

Comment:
`push_back` in `std::vector` typically doubles the capacity so that its not constantly re-allocating the buffer. The current implementation that uses `std::vector` might be doing this internally in the reserve calls, but if you switch to a `std::unique_ptr<char[]>` buffer that logic would have to be done here. Also strongly consider my recommendation to dropping the method entirely.